runtime/compiler/control/CompilationThread.cpp

```diff
@@ -7058,6 +7043,22 @@ TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary *portLib, void *
 
          TR_ASSERT(p->_optimizationPlan, "Must have an optimization plan");
 
+         bool aotCompilationReUpgradedToWarm = false;
+         if (that->_methodBeingCompiled->_useAotCompilation)
+            {
+            // In some circumstances AOT compilations are performed at warm
+            if ((TR::Options::getCmdLineOptions()->getAggressivityLevel() == TR::Options::AGGRESSIVE_AOT ||
+               that->getCompilationInfo()->importantMethodForStartup((J9Method*)method) ||
+               !TR::Options::getAOTCmdLineOptions()->getOption(TR_DisableAotAtCheapWarm)) &&
+               p->_optimizationPlan->isOptLevelDowngraded() &&
+               p->_optimizationPlan->getOptLevel() == cold) // Is this test really needed?
+               {
+               p->_optimizationPlan->setOptLevel(warm);
+               p->_optimizationPlan->setOptLevelDowngraded(false);
+               aotCompilationReUpgradedToWarm = true;
```

Comment:
Is this complication in the logic the best way to implement this ? 
I'm asking specifically about the opt level having been downgraded first (the condition that was checked before we entered this code) and then "reupgraded" afterwards ? i.e. it almost seems like we went through a warm -> cold -> warm transition in the decision making. Is it necessary to go through these hops to arrive at the same answer we started from, namely "warm". 
Maybe this is unavoidable given some other complexities of the code as it stands but I thought I would at least ask the question.

Reply:
I felt that we should not apply the cheap-warm strategy to methods that would not be subject to downgrading, i.e. if I decided that this method should stay at warm, then it should be a 'true' warm optimization level (in particular, if -Xjit:dontDowngradeToCold is used I don't want the switch to cheap-warm). Also, if the user specified -Xjit:initialOptlevel=cold (or even optLevel=cold) I don't want the push to cheap-warm.
Thus, first I need `shouldDowngradeOptLevel()` to run its course and arrive to a decision and only then override that decision. This overriding could have happened in `shouldDowngradeOptLevel()` but in this case I have to remember this decision until the options object is created later on. I had such a solution where I added another flag in OptimizationPlan and then I changed my mind because OptimizationPlan is defined in omr and I wanted to avoid 'polluting' the omr layer with openj9 specific flags

Reply:
Okay yes, I was expecting an answer that touched on the difference between "real" warm and "cheap" warm. Thanks, I understand the motivation for doing it this way.